probe/docker/registry.go

```diff
@@ -0,0 +1,273 @@
+package docker
+
+import (
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	docker_client "github.com/fsouza/go-dockerclient"
+)
+
+// Consts exported for testing.
+const (
+	StartEvent = "start"
+	DieEvent   = "die"
+	endpoint   = "unix:///var/run/docker.sock"
+)
+
+// Vars exported for testing.
+var (
+	NewDockerClientStub = newDockerClient
+	NewContainerStub    = NewContainer
+)
+
+// Registry keeps track of running docker containers and their images
+type Registry interface {
+	Stop()
+	LockedPIDLookup(f func(func(int) Container))
+	WalkContainers(f func(Container))
+	WalkImages(f func(*docker_client.APIImages))
+}
+
+type registry struct {
+	sync.RWMutex
+	quit     chan struct{}
+	interval time.Duration
+	client   Client
+
+	containers      map[string]Container
+	containersByPID map[int]Container
+	images          map[string]*docker_client.APIImages
+}
+
+// Client interface for mocking.
+type Client interface {
+	ListContainers(docker_client.ListContainersOptions) ([]docker_client.APIContainers, error)
+	InspectContainer(string) (*docker_client.Container, error)
+	ListImages(docker_client.ListImagesOptions) ([]docker_client.APIImages, error)
+	AddEventListener(chan<- *docker_client.APIEvents) error
+	RemoveEventListener(chan *docker_client.APIEvents) error
+}
+
+func newDockerClient(endpoint string) (Client, error) {
+	return docker_client.NewClient(endpoint)
+}
+
+// NewRegistry returns a usable Registry. Don't forget to Stop it.
+func NewRegistry(interval time.Duration) (Registry, error) {
+	client, err := NewDockerClientStub(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	r := &registry{
+		containers:      map[string]Container{},
+		containersByPID: map[int]Container{},
+		images:          map[string]*docker_client.APIImages{},
+
+		client:   client,
+		interval: interval,
+		quit:     make(chan struct{}),
+	}
+
+	go r.loop()
+	return r, nil
+}
+
+// Stop stops the Docker tagger's event subscriber.
+func (r *registry) Stop() {
+	close(r.quit)
+}
+
+func (r *registry) loop() {
+	if !r.update() {
+		return
+	}
+
+	ticker := time.Tick(r.interval)
+	for {
+		select {
+		case <-ticker:
+			if !r.update() {
+				return
+			}
+
+		case <-r.quit:
+			return
+		}
+	}
+}
+
+func (r *registry) update() bool {
```

Comment:
You're locking/unlocking the registry a bunch during this process, which could lead to queries returning half-complete data.